### PR TITLE
Lazy tensors & other memory changes

### PR DESCRIPTION
**Pull request by ndryden**
This primarily adds support for lazy allocation in tensors. This involved significant reworking of `RawBuffer` and `StridedMemory`.

### CODE REVIEW

File: include/h2/tensor/fixed_size_tuple.hpp [hunk 1]
```diff
@@ -199,6 +199,21 @@ constexpr AccT inner_product(const FixedSizeTuple<T1, SizeType1, N1>& a,
...
+template <typename T, typename SizeType, SizeType N, typename Predicate>
+constexpr bool any_of(const FixedSizeTuple<T, SizeType, N>& tuple,
+                      Predicate p) {
+  for (SizeType i = 0; i < tuple.size(); ++i) {
+    if (p(tuple[i])) {
+      return true;
+    }
+  }
+  return false;
+}
```
Comment (benson31) [inline include/h2/tensor/fixed_size_tuple.hpp:295]: I assume the reason you wrote this yourself is `std::any_of` will not be `constexpr` until 2020 (if/when that should ever happen...)?

File: include/h2/tensor/raw_buffer.hpp [hunk 2]
```diff
@@ -82,23 +82,41 @@ template <typename T, Device Dev>
 class RawBuffer {
 public:
   RawBuffer(const SyncInfo<Dev>& sync = SyncInfo<Dev>{})
-    : buffer(nullptr), buffer_size(0), sync_info(sync)
+    : buffer(nullptr), buffer_size(0), sync_info(sync), unowned_buffer(false)
   {}
-  RawBuffer(std::size_t size, const SyncInfo<Dev>& sync = SyncInfo<Dev>{})
-    : buffer(nullptr), buffer_size(size), sync_info(sync)
+  RawBuffer(std::size_t size, bool defer_alloc = false,
```
Comment (benson31) [inline include/h2/tensor/raw_buffer.hpp:87]: Not a fan of the "magic bool" in `defer_alloc`, but maybe this is low enough that it's ok.

File: include/h2/tensor/strided_memory.hpp [hunk 3]
```diff
@@ -204,18 +262,49 @@ class StridedMemory {
...
   SyncInfo<Dev> sync_info;  /**< Synchronization info for operations. */
+  bool is_mem_lazy;  /**< Whether allocation is lazy. */
+
+  /** Helper to create a raw buffer if size is non-empty. */
+  void make_raw_buffer(bool lazy)
+  {
+    // Do not allocate a RawBuffer for empty memory.
+    if (!mem_shape.empty())
+    {
+      std::size_t size = product<std::size_t>(mem_shape);
```
Comment (benson31) [inline include/h2/tensor/strided_memory.hpp]:
```suggestion
      std::size_t const size = product<std::size_t>(mem_shape);
```

File: include/h2/tensor/tensor.hpp [hunk 4]
```diff
@@ -29,13 +29,14 @@ class Tensor : public BaseTensor<T> {
...
+         bool lazy = false,
          const SyncInfo<Dev>& sync = SyncInfo<Dev>{}) :
     BaseTensor<T>(shape_, dim_types_),
-    tensor_memory(shape_, sync)
+    tensor_memory(shape_, lazy, sync)
   {}
 
-  Tensor(const SyncInfo<Dev>& sync = SyncInfo<Dev>{})
-    : Tensor(ShapeTuple(), DimensionTypeTuple(), sync) {}
+  Tensor(bool lazy = false, const SyncInfo<Dev>& sync = SyncInfo<Dev>{})
```
Comment (benson31) [inline include/h2/tensor/tensor.hpp]: At this level, we should definitely try to ameliorate the "magic bool" -- `Tensor foo(false, sync)` is almost certainly meaningless to a reader without extra commentary. I'm fine externalizing to a set of factory function overloads (`make_tensor(...)`, `make_lazy_tensor(...)`, etc) and leaving the ctors as-is, or perhaps tagging instead? Or something else?

**Review by benson31 — APPROVED** [review]

File: include/h2/tensor/fixed_size_tuple.hpp (hunk 1 above)
Comment (ndryden) [reply include/h2/tensor/fixed_size_tuple.hpp:295]: Yeah, though we should probably look at setting up an iterator-compatible interface for `FixedSizeTuple` anyway.

Comment (benson31) [reply include/h2/tensor/fixed_size_tuple.hpp:295]: This should be trivial as you can just return the relevant pointers for `begin` and `end`.

**Review by benson31 — APPROVED** [review]

**Review by benson31 — APPROVED** [review]
LGTM. Thanks for adding the iters and updating tests!

**Pull request merged**